staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go

```diff
@@ -513,3 +515,104 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
 func (u unionTransformers) TransformToStorage(ctx context.Context, data []byte, dataCtx value.Context) (out []byte, err error) {
 	return u[0].TransformToStorage(ctx, data, dataCtx)
 }
+
+// DynamicTransformers holds transformers that are dynamically updated.
+type DynamicTransformers struct {
+	transformTracker *atomic.Value
+	stopCh           <-chan struct{}
+}
+
+type transformTracker struct {
+	transformerOverrides map[schema.GroupResource]value.Transformer
+	closeTransformers    context.CancelFunc
+	inFlight             *atomic.Int64
+}
+
+// NewDynamicTransformers initializes DynamicTransformers when api-server starts.
+func NewDynamicTransformers(
+	transformerOverrides map[schema.GroupResource]value.Transformer,
+	closeTransformers context.CancelFunc,
+	stopCh <-chan struct{},
+) *DynamicTransformers {
+	dynamicTransformers := &DynamicTransformers{
+		transformTracker: &atomic.Value{},
+		stopCh:           stopCh,
+	}
+
+	tracker := &transformTracker{
+		transformerOverrides: transformerOverrides,
+		closeTransformers:    closeTransformers,
+		inFlight:             &atomic.Int64{},
+	}
+	dynamicTransformers.transformTracker.Store(tracker)
+
+	return dynamicTransformers
+}
+
+// TransformerForResource returns the transformer for the given resource.
+func (d *DynamicTransformers) TransformerForResource(resource schema.GroupResource) value.Transformer {
+	return &resourceTransformer{
+		resource:         resource,
+		transformTracker: d.transformTracker,
+	}
+}
+
+// Set sets the transformer overrides. This method is not go routine safe and must only be called by the same, single caller throughout the lifetime of this object.
+func (d *DynamicTransformers) Set(
+	transformerOverrides map[schema.GroupResource]value.Transformer,
+	closeTransformers context.CancelFunc,
+	inFlight *atomic.Int64,
+) {
+	tracker := d.transformTracker.Load().(*transformTracker)
+	oldCloseTransformers := tracker.closeTransformers
+	oldInFlight := tracker.inFlight
+
+	// store new values
+	newTransformTracker := &transformTracker{
+		transformerOverrides: transformerOverrides,
+		closeTransformers:    closeTransformers,
+		inFlight:             inFlight,
+	}
+	d.transformTracker.Store(newTransformTracker)
+
+	// poll until oldInFlight is zero.
+	go wait.PollImmediateUntil(1*time.Second, func() (bool, error) {
+		// check for oldInFlight to be zero to ensure they are not in use.
+		if oldInFlight.Load() == 0 {
+			oldCloseTransformers()
+			return true, nil
+		}
+		return false, nil
+	}, d.stopCh)
```

Comment:
Hmm if `oldInFlight` never reaches zero (say the transformer hangs forever), this would poll forever since this stop channel is only closed on process exit.  I think `.Set` should simply mark the tracker as `closing`, and in the `Transform` methods of `resourceTransformer` we should call `oldCloseTransformers` if `inFlight` `==` `0` and `closing==true`.  This removes the need for this `stopCh` and polling entirely.

After `closing`, `.Set` would probably need to check if `inFlight` `==` `0`, and then also call `oldCloseTransformers`.  In the `Transform` methods, if `closing==true` and `inFlight` `==` `0` on start, we would likely need to issue another `Load`.